lib/jxl/enc_frame.cc

```diff
@@ -1572,39 +1629,307 @@ bool CanDoStreamingEncoding(const CompressParams& cparams,
+  return true;
+}
+
 Status EncodeFrameStreaming(const CompressParams& cparams,
                             const FrameInfo& frame_info,
                             const CodecMetadata* metadata,
                             JxlEncoderChunkedFrameAdapter& frame_data,
                             const JxlCmsInterface& cms, ThreadPool* pool,
                             JxlEncoderOutputProcessorWrapper* output_processor,
                             AuxOut* aux_out) {
-  // TODO(szabadka): Call ComputeEncodingData for one DC group at a time and
-  // write the output at a precomputed position.
   PassesEncoderState enc_state;
+  SetProgressiveMode(cparams, &enc_state.progressive_splitter);
   FrameHeader frame_header(metadata);
-  std::vector<BitWriter> group_codes;
-  JXL_RETURN_IF_ERROR(
-      ComputeEncodingData(cparams, frame_info, metadata, frame_header,
-                          frame_data, 0, 0, frame_data.xsize, frame_data.ysize,
-                          cms, pool, enc_state, &group_codes, aux_out));
-
-  BitWriter writer;
-  writer.AppendByteAligned(enc_state.special_frames);
-  JXL_RETURN_IF_ERROR(WriteFrameHeader(frame_header, &writer, aux_out));
-
+  std::unique_ptr<jpeg::JPEGData> jpeg_data;
+  if (frame_data.IsJPEG()) {
+    jpeg_data = make_unique<jpeg::JPEGData>(frame_data.TakeJPEGData());
+  }
+  JXL_RETURN_IF_ERROR(MakeFrameHeader(frame_data.xsize, frame_data.ysize,
+                                      cparams, enc_state.progressive_splitter,
+                                      frame_info, jpeg_data.get(), true,
+                                      &frame_header));
   const size_t num_passes = enc_state.progressive_splitter.GetNumPasses();
+  ModularFrameEncoder enc_modular(frame_header, cparams);
   std::vector<coeff_order_t> permutation;
-  JXL_RETURN_IF_ERROR(PermuteGroups(cparams, enc_state.shared.frame_dim,
-                                    num_passes, &permutation, &group_codes));
-
-  JXL_RETURN_IF_ERROR(
-      WriteGroupOffsets(group_codes, permutation, &writer, aux_out));
-
-  writer.AppendByteAligned(group_codes);
-  PaddedBytes frame_bytes = std::move(writer).TakeBytes();
-  JXL_RETURN_IF_ERROR(AppendData(*output_processor, frame_bytes));
-
+  std::vector<size_t> dc_group_order;
+  ComputePermutationForStreaming(frame_data.xsize, frame_data.ysize, num_passes,
+                                 permutation, dc_group_order);
+  enc_state.shared.num_histograms = dc_group_order.size();
+  // This is only valid in VarDCT mode, otherwise there can be group shift.
+  size_t group_size = 256;
+  size_t dc_group_size = group_size * kBlockDim;
+  size_t dc_group_xsize = DivCeil(frame_data.xsize, dc_group_size);
+  size_t min_dc_global_size = 0;
+  size_t group_data_offset = 0;
+  PaddedBytes frame_header_bytes;
+  PaddedBytes dc_global_bytes;
+  std::vector<size_t> group_sizes;
+  size_t start_pos = output_processor->CurrentPosition();
+  for (size_t i = 0; i < dc_group_order.size(); ++i) {
```

Comment:
This logic is going to be an absolute pleasure to parallelize......

Reply:
Yes, one step at a time :)